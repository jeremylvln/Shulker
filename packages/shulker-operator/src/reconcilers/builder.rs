use std::fmt::Debug;
use tracing::*;

use kube::{
    api::{DeleteParams, Patch, PatchParams, PostParams},
    Api, ResourceExt,
};
use serde::{de::DeserializeOwned, Serialize};

#[async_trait::async_trait]
pub trait ResourceBuilder {
    type OwnerType: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug;
    type ResourceType: kube::Resource<DynamicType = ()>
        + Clone
        + Serialize
        + DeserializeOwned
        + Debug;

    fn name(owner: &Self::OwnerType) -> String;
    fn api(&self, owner: &Self::OwnerType) -> Api<Self::ResourceType>;

    fn is_needed(&self, _owner: &Self::OwnerType) -> bool {
        true
    }

    fn is_recreation_needed(
        _owner: &Self::OwnerType,
        _existing_resource: &Self::ResourceType,
    ) -> bool {
        false
    }

    async fn build(
        &self,
        owner: &Self::OwnerType,
        name: &str,
        existing_resource: Option<&Self::ResourceType>,
    ) -> Result<Self::ResourceType, anyhow::Error>;
}

async fn get_existing<
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
>(
    api: &Api<R>,
    name: &str,
) -> Result<Option<R>, anyhow::Error> {
    api.get(name).await.map(|r| Some(r)).or_else(|e| match e {
        kube::Error::Api(e) => match e.code {
            404 => Ok(None),
            _ => Err(e.into()),
        },
        e => Err(e.into()),
    })
}

fn set_controller_reference<
    O: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
>(
    owner: &O,
    resource: &mut R,
) {
    resource
        .owner_references_mut()
        .push(owner.controller_owner_ref(&()).unwrap());
}

pub async fn reconcile_builder<
    O: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    RB: ResourceBuilder<OwnerType = O, ResourceType = R>,
>(
    builder: &RB,
    owner: &O,
) -> super::Result<Option<R>> {
    let api = builder.api(owner);
    let name = RB::name(owner);

    debug!(
        builder = std::any::type_name::<RB>(),
        owner = std::any::type_name::<O>(),
        owner_name = owner.name_any(),
        target = std::any::type_name::<R>(),
        target_name = name,
        "reconciling builder",
    );

    let mut existing_resource = get_existing(&api, &name)
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e))?;

    if existing_resource.is_some() {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "found existing resource",
        );

        let delete_existing = || async {
            api.delete(&name, &DeleteParams::default())
                .await
                .map_err(|e| {
                    super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into())
                })
        };

        if !builder.is_needed(owner) {
            debug!(
                builder = std::any::type_name::<RB>(),
                owner = std::any::type_name::<O>(),
                owner_name = owner.name_any(),
                target = std::any::type_name::<R>(),
                target_name = name,
                "existing resource is not needed anymore, deleting",
            );

            delete_existing().await?;
            return Ok(None);
        } else if RB::is_recreation_needed(owner, existing_resource.as_ref().unwrap()) {
            debug!(
                builder = std::any::type_name::<RB>(),
                owner = std::any::type_name::<O>(),
                owner_name = owner.name_any(),
                target = std::any::type_name::<R>(),
                target_name = name,
                "existing resource needs to be recreated, deleting",
            );

            delete_existing().await?;
            existing_resource = None;
        }
    }

    if !builder.is_needed(owner) {
        return Ok(None);
    }

    let mut new_resource = builder
        .build(owner, &name, existing_resource.as_ref())
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e))?;

    let updated_resource = if let Some(existing_resource) = existing_resource {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "patching existing resource",
        );

        for (key, value) in existing_resource.annotations().iter() {
            if !new_resource.annotations().contains_key(key) {
                new_resource
                    .annotations_mut()
                    .insert(key.clone(), value.clone());
            }
        }

        api.patch(
            &name,
            &PatchParams::apply("shulker-operator").force(),
            &Patch::Apply(&new_resource),
        )
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into()))?
    } else {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "creating new resource",
        );

        set_controller_reference(owner, &mut new_resource);
        api.create(
            &PostParams {
                dry_run: false,
                field_manager: Some("shulker-operator".to_string()),
            },
            &new_resource,
        )
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into()))?
    };

    Ok(Some(updated_resource))
}
