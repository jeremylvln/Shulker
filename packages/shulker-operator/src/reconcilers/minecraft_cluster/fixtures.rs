use http::{Request, Response};
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::v1alpha1::minecraft_cluster::{
    MinecraftCluster, MinecraftClusterExternalServerSpec, MinecraftClusterSpec,
};

lazy_static! {
    pub static ref TEST_CLUSTER: MinecraftCluster = MinecraftCluster {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-cluster".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftClusterSpec {
            network_admins: None,
            redis: None,
            external_servers: Some(vec![MinecraftClusterExternalServerSpec {
                name: "my-external-server".to_string(),
                address: "127.0.0.1:25565".to_string(),
                tags: vec!["game".to_string()]
            }])
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
