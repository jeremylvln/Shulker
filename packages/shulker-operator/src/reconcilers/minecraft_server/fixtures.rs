use std::collections::BTreeMap;

use http::{Request, Response};
use k8s_openapi::api::core::v1::EnvVar;
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        minecraft_server::{
            MinecraftServer, MinecraftServerConfigurationLifecycleStrategy,
            MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
            MinecraftServerPodOverridesSpec, MinecraftServerSpec, MinecraftServerVersion,
            MinecraftServerVersionSpec,
        },
    },
};

lazy_static! {
    pub static ref TEST_SERVER: MinecraftServer = MinecraftServer {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-server".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftServerSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            tags: vec!["lobby".to_string()],
            version: MinecraftServerVersionSpec {
                channel: MinecraftServerVersion::Paper,
                name: "1.20.1".to_string(),
                ..MinecraftServerVersionSpec::default()
            },
            config: MinecraftServerConfigurationSpec {
                existing_config_map_name: None,
                world: Some(ResourceRefSpec {
                    url: Some("https://example.com/my_world.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }),
                plugins: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_plugin.jar".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                patches: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_patch.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                max_players: 42,
                disable_nether: false,
                disable_end: true,
                server_properties: None,
                proxy_forwarding_mode: MinecraftServerConfigurationProxyForwardingMode::Velocity,
                lifecycle_strategy:
                    MinecraftServerConfigurationLifecycleStrategy::AllocateWhenNotEmpty,
                skip_agent_download: false
            },
            pod_overrides: Some(MinecraftServerPodOverridesSpec {
                image: None,
                env: Some(vec![EnvVar {
                    name: "EXTRA_ENV".to_string(),
                    value: Some("my_value".to_string()),
                    ..EnvVar::default()
                }]),
                resources: None,
                affinity: None,
                node_selector: Some(BTreeMap::from([(
                    "beta.kubernetes.io/os".to_string(),
                    "linux".to_string()
                )])),
                tolerations: None,
                service_account_name: None,
                volume_mounts: None,
                volumes: None,
                ports: None,
                health: None,
            })
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
