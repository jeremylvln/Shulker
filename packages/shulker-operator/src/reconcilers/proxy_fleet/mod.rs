use std::{collections::BTreeMap, sync::Arc, time::Duration};

use futures::StreamExt;
use google_agones_crds::v1::{fleet::Fleet, fleet_autoscaler::FleetAutoscaler};
use k8s_openapi::api::core::v1::{ConfigMap, Service};
use kube::{
    api::{ListParams, PatchParams},
    runtime::{controller::Action, watcher::Config, Controller},
    Api, Client, ResourceExt,
};
use tracing::*;

use shulker_crds::{
    condition::{ConditionStatus, HasConditions},
    v1alpha1::proxy_fleet::{ProxyFleet, ProxyFleetStatus},
};

use self::{
    config_map::ConfigMapBuilder, fleet::FleetBuilder, fleet_autoscaler::FleetAutoscalerBuilder,
    service::ServiceBuilder,
};

use super::{builder::reconcile_builder, status::patch_status, ReconcilerError, Result};

mod config_map;
mod fleet;
mod fleet_autoscaler;
mod service;

#[cfg(test)]
mod fixtures;

struct ProxyFleetReconciler {
    client: kube::Client,

    // Builders
    config_map_builder: ConfigMapBuilder,
    service_builder: ServiceBuilder,
    fleet_builder: FleetBuilder,
    fleet_autoscaler_builder: FleetAutoscalerBuilder,
}

impl ProxyFleetReconciler {
    async fn reconcile(
        &self,
        api: Api<ProxyFleet>,
        proxy_fleet: Arc<ProxyFleet>,
    ) -> Result<Action> {
        reconcile_builder(&self.config_map_builder, proxy_fleet.as_ref()).await?;
        reconcile_builder(&self.service_builder, proxy_fleet.as_ref()).await?;
        let fleet = reconcile_builder(&self.fleet_builder, proxy_fleet.as_ref()).await?;
        reconcile_builder(&self.fleet_autoscaler_builder, proxy_fleet.as_ref()).await?;

        if let Some(fleet) = &fleet {
            if let Some(fleet_status) = &fleet.status {
                let mut proxy_fleet = proxy_fleet.as_ref().clone();
                if proxy_fleet.status.is_none() {
                    proxy_fleet.status = Some(ProxyFleetStatus::default());
                }

                let status = proxy_fleet.status.as_mut().unwrap();

                status.replicas = fleet_status.replicas;
                status.ready_replicas = fleet_status.ready_replicas;
                status.allocated_replicas = fleet_status.allocated_replicas;

                if status.ready_replicas > 0 || status.allocated_replicas > 0 {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::True,
                        "AtLeastOneReadyOrAllocated".to_string(),
                        "One or more servers are ready or allocated".to_string(),
                    );
                } else {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::False,
                        "NoneReady".to_string(),
                        "No server are ready".to_string(),
                    );
                };

                patch_status(
                    &api,
                    &PatchParams::apply("shulker-operator").force(),
                    &proxy_fleet,
                )
                .await?;
            }
        }

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, proxy_fleet: Arc<ProxyFleet>) -> Result<Action> {
        info!(
            name = proxy_fleet.name_any(),
            namespace = proxy_fleet.namespace(),
            "cleaning up ProxyFleet",
        );

        Ok(Action::await_change())
    }

    fn get_labels(
        proxy_fleet: &ProxyFleet,
        name: String,
        component: String,
    ) -> BTreeMap<String, String> {
        BTreeMap::from([
            ("app.kubernetes.io/name".to_string(), name.clone()),
            (
                "app.kubernetes.io/instance".to_string(),
                format!("{}-{}", name, proxy_fleet.name_any()),
            ),
            ("app.kubernetes.io/component".to_string(), component),
            (
                "app.kubernetes.io/part-of".to_string(),
                format!("cluster-{}", proxy_fleet.spec.cluster_ref.name.clone()),
            ),
            (
                "app.kubernetes.io/managed-by".to_string(),
                "shulker-operator".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                proxy_fleet.spec.cluster_ref.name.clone(),
            ),
            (
                "proxyfleet.shulkermc.io/name".to_string(),
                proxy_fleet.name_any(),
            ),
        ])
    }
}

#[instrument(skip(ctx, proxy_fleet))]
async fn reconcile(proxy_fleet: Arc<ProxyFleet>, ctx: Arc<ProxyFleetReconciler>) -> Result<Action> {
    let ns = proxy_fleet.namespace().unwrap();
    let proxy_fleets_api: Api<ProxyFleet> = Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = proxy_fleet.name_any(),
        namespace = ns,
        "reconciling ProxyFleet",
    );

    if proxy_fleet.metadata.deletion_timestamp.is_none() {
        ctx.reconcile(proxy_fleets_api.clone(), proxy_fleet.clone())
            .await
    } else {
        ctx.cleanup(proxy_fleet.clone()).await
    }
}

fn error_policy(
    _proxy_fleet: Arc<ProxyFleet>,
    error: &ReconcilerError,
    _ctx: Arc<ProxyFleetReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&proxy_fleet, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client) {
    let proxy_fleets_api = Api::<ProxyFleet>::all(client.clone());
    if let Err(e) = proxy_fleets_api.list(&ListParams::default().limit(1)).await {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = ProxyFleetReconciler {
        client: client.clone(),
        config_map_builder: ConfigMapBuilder::new(client.clone()),
        service_builder: ServiceBuilder::new(client.clone()),
        fleet_builder: FleetBuilder::new(client.clone()),
        fleet_autoscaler_builder: FleetAutoscalerBuilder::new(client.clone()),
    };

    Controller::new(proxy_fleets_api, Config::default().any_semantic())
        .owns(
            Api::<ConfigMap>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Service>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Fleet>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<FleetAutoscaler>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .shutdown_on_signal()
        .run(reconcile, error_policy, context.into())
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;
}
