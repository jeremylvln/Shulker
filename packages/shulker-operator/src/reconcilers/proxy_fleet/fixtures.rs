use std::collections::BTreeMap;

use google_agones_crds::v1::fleet_autoscaler::{
    FleetAutoscalerPolicy, FleetAutoscalerPolicyBufferSpec, FleetAutoscalerPolicySpec,
};
use http::{Request, Response};
use k8s_openapi::{api::core::v1::EnvVar, apimachinery::pkg::util::intstr::IntOrString};
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    schemas::{FleetAutoscalingSpec, TemplateSpec},
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        proxy_fleet::{
            ProxyFleet, ProxyFleetServiceExternalTrafficPolicy, ProxyFleetServiceSpec,
            ProxyFleetServiceType, ProxyFleetSpec, ProxyFleetTemplateConfigurationSpec,
            ProxyFleetTemplatePodOverridesSpec, ProxyFleetTemplateSpec, ProxyFleetTemplateVersion,
            ProxyFleetTemplateVersionSpec,
        },
    },
};

lazy_static! {
    pub static ref TEST_PROXY_FLEET: ProxyFleet = ProxyFleet {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-proxy".to_string()),
            ..ObjectMeta::default()
        },
        spec: ProxyFleetSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            replicas: 3,
            template: TemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(BTreeMap::from([(
                        "test-label/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    annotations: Some(BTreeMap::from([(
                        "test-annotation/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    ..ObjectMeta::default()
                }),
                spec: ProxyFleetTemplateSpec {
                    version: ProxyFleetTemplateVersionSpec {
                        channel: ProxyFleetTemplateVersion::Velocity,
                        name: "latest".to_string(),
                    },
                    config: ProxyFleetTemplateConfigurationSpec {
                        existing_config_map_name: None,
                        plugins: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_plugin.jar".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        patches: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_patch.tar.gz".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        max_players: 1000,
                        motd: "A Motd".to_string(),
                        server_icon: "abc==".to_string(),
                        proxy_protocol: true,
                        ttl_seconds: 3600,
                        players_delta_before_exclusion: 15,
                        skip_agent_download: false
                    },
                    pod_overrides: Some(ProxyFleetTemplatePodOverridesSpec {
                        image: None,
                        env: Some(vec![EnvVar {
                            name: "EXTRA_ENV".to_string(),
                            value: Some("my_value".to_string()),
                            ..EnvVar::default()
                        }]),
                        resources: None,
                        affinity: None,
                        node_selector: Some(BTreeMap::from([(
                            "beta.kubernetes.io/os".to_string(),
                            "linux".to_string()
                        )])),
                        tolerations: None,
                        service_account_name: None,
                        volume_mounts: None,
                        volumes: None,
                        ports: None,
                    })
                },
            },
            service: Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::LoadBalancer,
                annotations: Some(BTreeMap::from([(
                    "service.beta.kubernetes.io/load-balancer".to_string(),
                    "internal".to_string()
                )])),
                external_traffic_policy: Some(ProxyFleetServiceExternalTrafficPolicy::Cluster),
                preferred_reconnection_address: Some("127.0.0.1".to_string())
            }),
            autoscaling: Some(FleetAutoscalingSpec {
                agones_policy: Some(FleetAutoscalerPolicySpec {
                    type_: FleetAutoscalerPolicy::Buffer,
                    buffer: Some(FleetAutoscalerPolicyBufferSpec {
                        max_replicas: 8,
                        min_replicas: 3,
                        buffer_size: IntOrString::Int(2),
                    }),
                    webhook: None,
                })
            }),
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
