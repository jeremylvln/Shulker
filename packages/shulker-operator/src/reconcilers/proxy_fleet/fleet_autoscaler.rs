use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use google_agones_crds::v1::fleet_autoscaler::FleetAutoscaler;
use google_agones_crds::v1::fleet_autoscaler::FleetAutoscalerSpec;
use shulker_crds::v1alpha1::proxy_fleet::ProxyFleet;

use super::fleet::FleetBuilder;
use super::ProxyFleetReconciler;

pub struct FleetAutoscalerBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for FleetAutoscalerBuilder {
    type OwnerType = ProxyFleet;
    type ResourceType = FleetAutoscaler;

    fn name(proxy_fleet: &Self::OwnerType) -> String {
        proxy_fleet.name_any()
    }

    fn api(&self, proxy_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            proxy_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, proxy_fleet: &Self::OwnerType) -> bool {
        match &proxy_fleet.spec.autoscaling {
            Some(autoscaling) => autoscaling.agones_policy.is_some(),
            None => false,
        }
    }

    async fn build(
        &self,
        proxy_fleet: &Self::OwnerType,
        name: &str,
        _existing_fleet_autoscaler: Option<&Self::ResourceType>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let fleet_autoscaler = FleetAutoscaler {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(proxy_fleet.namespace().unwrap().clone()),
                labels: Some(
                    ProxyFleetReconciler::get_common_labels(proxy_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            spec: FleetAutoscalerSpec {
                fleet_name: FleetBuilder::name(proxy_fleet),
                policy: proxy_fleet
                    .spec
                    .autoscaling
                    .as_ref()
                    .unwrap()
                    .agones_policy
                    .as_ref()
                    .unwrap()
                    .clone(),
            },
            status: None,
        };

        Ok(fleet_autoscaler)
    }
}

impl FleetAutoscalerBuilder {
    pub fn new(client: Client) -> Self {
        FleetAutoscalerBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        proxy_fleet::fixtures::{create_client_mock, TEST_PROXY_FLEET},
    };

    #[test]
    fn name_contains_fleet_name() {
        // W
        let name = super::FleetAutoscalerBuilder::name(&TEST_PROXY_FLEET);

        // T
        assert_eq!(name, "my-proxy");
    }

    #[tokio::test]
    async fn is_needed_with_autoscaling() {
        // G
        let client = create_client_mock();
        let builder = super::FleetAutoscalerBuilder::new(client);

        // W
        let is_needed = builder.is_needed(&TEST_PROXY_FLEET);

        // T
        assert!(is_needed);
    }

    #[tokio::test]
    async fn is_needed_without_autoscaling() {
        // G
        let client = create_client_mock();
        let builder = super::FleetAutoscalerBuilder::new(client);
        let mut fleet = TEST_PROXY_FLEET.clone();
        fleet.spec.autoscaling = None;

        // W
        let is_needed = builder.is_needed(&fleet);

        // T
        assert!(!is_needed);
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::FleetAutoscalerBuilder::new(client);
        let name = super::FleetAutoscalerBuilder::name(&TEST_PROXY_FLEET);

        // W
        let fleet_autoscaler = builder.build(&TEST_PROXY_FLEET, &name, None).await.unwrap();

        // T
        insta::assert_yaml_snapshot!(fleet_autoscaler);
    }
}
