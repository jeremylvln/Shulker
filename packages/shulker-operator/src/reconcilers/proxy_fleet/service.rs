use k8s_openapi::api::core::v1::Service;
use k8s_openapi::api::core::v1::ServicePort;
use k8s_openapi::api::core::v1::ServiceSpec;
use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use shulker_crds::v1alpha1::proxy_fleet::ProxyFleet;
use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

use super::ProxyFleetReconciler;

pub struct ServiceBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for ServiceBuilder {
    type OwnerType = ProxyFleet;
    type ResourceType = Service;
    type Context = ();

    fn name(proxy_fleet: &Self::OwnerType) -> String {
        proxy_fleet.name_any()
    }

    fn api(&self, proxy_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            proxy_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, proxy_fleet: &Self::OwnerType) -> bool {
        proxy_fleet.spec.service.is_some()
    }

    async fn build(
        &self,
        proxy_fleet: &Self::OwnerType,
        name: &str,
        _existing_service: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let service_config = proxy_fleet.spec.service.as_ref().unwrap();
        let service = Service {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(proxy_fleet.namespace().unwrap().clone()),
                labels: Some(ProxyFleetReconciler::get_labels(
                    proxy_fleet,
                    "proxy".to_string(),
                    "proxy".to_string(),
                )),
                annotations: service_config.annotations.clone(),
                ..ObjectMeta::default()
            },
            spec: Some(ServiceSpec {
                selector: Some(ProxyFleetReconciler::get_labels(
                    proxy_fleet,
                    "proxy".to_string(),
                    "proxy".to_string(),
                )),
                type_: Some(service_config.type_.to_string()),
                external_traffic_policy: service_config
                    .external_traffic_policy
                    .as_ref()
                    .map(|x| x.to_string()),
                ports: Some(vec![ServicePort {
                    name: Some("minecraft".to_string()),
                    protocol: Some("TCP".to_string()),
                    port: 25565,
                    target_port: Some(IntOrString::String("minecraft".to_string())),
                    ..ServicePort::default()
                }]),
                ..ServiceSpec::default()
            }),
            ..Service::default()
        };

        Ok(service)
    }
}

impl ServiceBuilder {
    pub fn new(client: Client) -> Self {
        ServiceBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

    use crate::reconcilers::proxy_fleet::fixtures::{create_client_mock, TEST_PROXY_FLEET};

    #[test]
    fn name_contains_fleet_name() {
        // W
        let name = super::ServiceBuilder::name(&TEST_PROXY_FLEET);

        // T
        assert_eq!(name, "my-proxy");
    }

    #[tokio::test]
    async fn is_needed_with_service() {
        // G
        let client = create_client_mock();
        let builder = super::ServiceBuilder::new(client);

        // W
        let is_needed = builder.is_needed(&TEST_PROXY_FLEET);

        // T
        assert!(is_needed);
    }

    #[tokio::test]
    async fn is_needed_without_service() {
        // G
        let client = create_client_mock();
        let builder = super::ServiceBuilder::new(client);
        let mut fleet = TEST_PROXY_FLEET.clone();
        fleet.spec.service = None;

        // W
        let is_needed = builder.is_needed(&fleet);

        // T
        assert!(!is_needed);
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ServiceBuilder::new(client);
        let name = super::ServiceBuilder::name(&TEST_PROXY_FLEET);

        // W
        let service = builder
            .build(&TEST_PROXY_FLEET, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(service);
    }

    #[tokio::test]
    async fn build_add_custom_annotations() {
        // G
        let client = create_client_mock();
        let builder = super::ServiceBuilder::new(client);
        let custom_annotations = BTreeMap::from([(
            "service.beta.kubernetes.io/load-balancer".to_string(),
            "internal".to_string(),
        )]);
        let mut proxy_fleet = TEST_PROXY_FLEET.clone();
        proxy_fleet.spec.service.as_mut().unwrap().annotations = Some(custom_annotations.clone());
        let name = super::ServiceBuilder::name(&proxy_fleet);

        // W
        let service = builder
            .build(&proxy_fleet, &name, None, None)
            .await
            .unwrap();

        // T
        assert_eq!(service.metadata.annotations.unwrap(), custom_annotations);
    }
}
