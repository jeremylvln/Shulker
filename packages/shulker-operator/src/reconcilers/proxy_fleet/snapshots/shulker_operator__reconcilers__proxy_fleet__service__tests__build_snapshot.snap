---
source: packages/shulker-operator/src/reconcilers/proxy_fleet/service.rs
expression: service
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/load-balancer: internal
  labels:
    app.kubernetes.io/component: proxy
    app.kubernetes.io/instance: proxy-my-proxy
    app.kubernetes.io/managed-by: shulker-operator
    app.kubernetes.io/name: proxy
    app.kubernetes.io/part-of: cluster-my-cluster
    minecraftcluster.shulkermc.io/name: my-cluster
    proxyfleet.shulkermc.io/name: my-proxy
  name: my-proxy
  namespace: default
spec:
  externalTrafficPolicy: Cluster
  ports:
    - name: minecraft
      port: 25565
      protocol: TCP
      targetPort: minecraft
  selector:
    app.kubernetes.io/component: proxy
    app.kubernetes.io/instance: proxy-my-proxy
    app.kubernetes.io/managed-by: shulker-operator
    app.kubernetes.io/name: proxy
    app.kubernetes.io/part-of: cluster-my-cluster
    minecraftcluster.shulkermc.io/name: my-cluster
    proxyfleet.shulkermc.io/name: my-proxy
  type: LoadBalancer

